### PR TITLE
logger/gelf: use compression level 0 by default

### PR DESCRIPTION
**Pull request by kolyshkin**
go-gelf package used by dockerd uses gzip compression by default,
which is costly in terms of CPU. Since the data is sent over the
network, it's a trade-off between CPU and network bandwidth.
    
This comes from observing the internal infrastructure on which
dockerd writes containers' logs to one of its containers
(which itself does something else with those logs). With
default gzip -1 compression, dockerd was using 150-220% CPU,
which decreased to about 30% without compression.

In addition, there are multiple reports of gelf using
too much CPU:
    
* https://github.com/moby/moby/issues/19665
* https://github.com/moby/moby/issues/19209
    
With this commit, we disable compression but the output
produced is still gzip-compatible. This is done because
there are reports that Logstash only got support for
uncompressed log messages recently (see https://github.com/logstash-plugins/logstash-input-gelf/pull/48).

**- Description for the changelog**

* log/gelf: default compression level set to 0 to prioritize saving CPU over network bandwidth

### CODE REVIEW

**Comment by kolyshkin** [comment]
Related doc change: https://github.com/docker/docker.github.io/pull/9678

**Comment by kolyshkin** [comment]
@mariussturm please let us know what do you think about this one

**Comment by mariussturm** [comment]
@kolyshkin do you have some numbers of the network utilization as well? 30% less CPU is nice but if the network is at its limit we don't win much.

**Comment by thaJeztah** [comment]
@mariussturm IIRC, it was not 30% _less_ but a drop from 150-220% CPU _to_ 30% CPU. Don't know if there's numbers about network utilisation though

**Comment by thaJeztah** [comment]
Hm. no idea why this test is failing (think it's a flaky swarm test)

```
=== RUN TestInspect
--- FAIL: TestInspect (11.94s)
inspect_test.go:23: Creating a new daemon at: "/go/src/github.com/docker/docker/bundles/test-integration/TestInspect"
inspect_test.go:39: timeout hit after 10s: waiting for tasks to enter run state. task failed with error: task: non-zero exit (1)
```

**Comment by kolyshkin** [comment]
> 30% less CPU is nice

@mariussturm In my case it was more like 200% less CPU. As for network utilization, well, in this particular case the receiving side was a container on the same host, so compression was totally useless.

Here is what I think:
- if someone sees high dockerd cpu utilization, it is usually very hard to figure out what is the root cause;
- if someone sees high network utilization, it is relatively easy to figure out what's causing it, as you have source/destination ip/port for every packet.

In other words, it is much harder to go from "dockerd eats lots of cpu" to "let's disable gelf compression" than from "network is saturated" to "let's enable gelf compression".

One other thing is, gzip is relatively CPU heavy, and not really good for streaming data.

**Comment by mariussturm** [comment]
@kolyshkin there is no need to convince me, you asked me for my opinion. My opinion is: You can only make a good decision when you compare CPU vs. Network utilization. When you have ten other arguments but you don't look at the relevant data just merge it and it's all fine.

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by kolyshkin** [comment]
> My opinion is: You can only make a good decision when you compare CPU vs. Network utilization. When you have ten other arguments but you don't look at the relevant data just merge it and it's all fine.

@mariussturm yes, the decision should be made on a case-by-case basis, looking into CPU and network utilization, which one is the bottleneck in this particular setup. Currently, dockerd provides an option for that, so a user can enable/disable compression based on their findings.

My question to you was a bit different: what should be the _default_ setting for that? Yes, it's a tricky question, considering we don't know the details of a particular setup, targeting the common case (whatever that might be) instead.

I am not trying to convince you here, it's more like thought-provoking -- "here's what I think and why, now what do you think and why?".

You had a question and an observation which I answered; let me quote and elaborate on both points.

>> 30% less CPU is nice

> In my case it was more like 200% less CPU

Let me elaborate. In the original changelog message I wrote "dockerd was using 150-220% CPU,
which decreased to about 30% without compression", which amounts to 120-190% less CPU, not 30% less.

>> do you have some numbers of the network utilization as well?

> As for network utilization, well, in this particular case the receiving side was a container on the same host

Let me elaborate. I believe that since the network traffic is going via loopback, any measurements here are irrelevant. Yet let me take a look. Note: it's a production system so it's not very reproducible so this is what I am seeing now on a system with default compression (gzip -1), with hardware being 8 core Intel Xeon E5-2666:

* ~3 mbps (megabits per second) average TX traffic on the gelf port;
* ~130% CPU usage by dockerd (of which I estimate 100% to be gzip-related).

Unfortunately this is a production system so making and changes are complicated and requires lots of coordination, so I can't do another set of measurements with no compression.

My rough guess you can expect 2x to 3x more traffic with compression disabled, so let's assume it would be 3x traffic, so about 10 mbps. I understand this is not very scientific as it involves some guessing, so YMMV.

Still, given the average common case (logs going to a machine in the same rack/datacenter, rather than overseas via a slow satellite link), I think that saving 6-7 mbps network bandwidth is worth is not worth 100% CPU utilization increase.

Again, I am not trying to convince you, just sincerely asking for help in finding out which default is better.

**Comment by mariussturm** [comment]
There you are, generating traffic a fraction of the max network bandwidth is totally not worth spending a whole CPU core. So the answer is pretty clear. The none compression default is better in this scenario.

**Comment by ingshtrom** [comment]
@kolyshkin I think this is more of a change than you might expect. We are using logstash:6.2.2 and if this was made in our environment today, we would no longer have logs for any of our applications, as was the case when we disabled compression in our stage environment. Based on @binman-docker's investigation, support for an uncompressed GELF message via Logstash is a recent addition.

All I'm saying is that believe Logstash is pretty heavily used and for any of those users using it with Docker their applications logs would stop. The documentation is great, but this should likely be documented as a breaking change in our release notes as well. I'm not sure how release notes are handled, though.

**Comment by thaJeztah** [comment]
could we change the default compression level; would that reduce the CPU load, while still being backward compatible?

**Comment by thaJeztah** [comment]
nevermind; looks like it defaults to `1` already 🤷‍♂️

**Comment by thaJeztah** [comment]
Actually; this looks iffy; https://github.com/moby/moby/blob/9a2c26374393e0f46e8f2b1e49d2dcf66b78dbb4/daemon/logger/gelf/gelf.go#L218

Values lower than "default" are marked invalid, but "default" is `-1`: https://golang.org/pkg/compress/flate/#pkg-constants

```go
const (
    NoCompression      = 0
    BestSpeed          = 1
    BestCompression    = 9
    DefaultCompression = -1

    // HuffmanOnly disables Lempel-Ziv match searching and only performs Huffman
    // entropy encoding. This mode is useful in compressing data that has
    // already been compressed with an LZ style algorithm (e.g. Snappy or LZ4)
    // that lacks an entropy encoder. Compression gains are achieved when
    // certain bytes in the input stream occur more frequently than others.
    //
    // Note that HuffmanOnly produces a compressed output that is
    // RFC 1951 compliant. That is, any valid DEFLATE decompressor will
    // continue to be able to decompress this output.
    HuffmanOnly = -2
)

**Comment by thaJeztah** [comment]
Nevermind; I guess `HuffmanOnly` may not be supported.

We should check what `DefaultCompression` is though (the docs mention it's `1` which may not be correct?

**Comment by kolyshkin** [comment]
Well, default compression is level 6 and not level 1:

```go
case level == DefaultCompression:
	level = 6
```

I also found out that pkg/deflate has a separate "fast" implementation that is used for `level==1`.

So, we should at least set the default to be level 1.

**Comment by kolyshkin** [comment]
Hmm, actually, `compress/flate` is only used by `gelf` for compression level constants; for the actual compression, either `compress/gzip` or `compress/zlib` is used. Default is `gzip`, and `gzip` apparently uses `flate`, and `gelf` sets the level to `1` by default, so as far as I understand this fast flate algo is actually used.

Even with all this, it still takes lots of CPU.

So, I think the best course of action would be to disable compression.

**Comment by kolyshkin** [comment]
@ingshtrom the alternative is to set compression level to 0. In this case, no compression is performed but the data is still wrapped in gzip format.

Is it possible that you test that change?

**Comment by ingshtrom** [comment]
you probably already have your answer via the private repository discussion, but I don't believe we can test this now since we've already moved to a logstash version that supports gelf without compression and have compression fully turned off.

**Comment by kolyshkin** [comment]
So, the "Compression: none" only works since logstash-input-gelf 3.3.0 (https://rubygems.org/gems/logstash-input-gelf/versions/3.3.0), which is included in logstash 7.4.0, released Oct 1, 2019 (https://www.elastic.co/guide/en/logstash/7.4/logstash-7-4-0.html).

This means we can't disable compression now since I'm afraid many users still haven't updated logstash to >= 7.4.

Now we need to check whether setting "compression-level: 0" works with older logstash

**Comment by kolyshkin** [comment]
Took a while to figure out how to test it. Here's how:

```
LOGSTASH_VERSION=7.3.1
docker run --rm -it -p 127.0.0.1:12201:12201/udp docker.elastic.co/logstash/logstash:${LOGSTASH_VERSION} bin/logstash -e 'input { gelf {} }'
# in another terminal window
docker run --rm --log-driver=gelf --log-opt=gelf-address=udp://127.0.0.1:12201 --log-opt=gelf-compression-type=gzip --log-opt=gelf-compression-level=0 alpine echo hahaha
```

Using the above test, found out that
* `gelf-compression-type=none` is not working with logstash < 7.4.0 (7.3.1 was tested)
* `gelf-compression-type=none` works with logstash >= 7.4.0 (7.4.2 was tested)
* `gelf-compression-type=gzip --log-opt=gelf-compression-level=0` works with all versions (7.3.1 and 7.4.2 were tested)

**Comment by kolyshkin** [comment]
OK, some more testing showed that using `gelf-compression-level: 0` is not making things better, dockerd CPU usage stays about the same, and the CPU profiles collected with `pprof` does not differ much from each other. Changing `gelf-compression-type` to `none` changes things dramatically (as described in earlier comments).

Apparently that happens since most of the CPU time is spent not on compression itself, but rather go runtime (allocation and garbage collection). Looks like gzipping a lot of small data objects is inefficient.

Given the fact that the support for uncompressed input has only made its way to logstash very recently (see https://github.com/moby/moby/pull/40101#issuecomment-553047340), we can't change the default to be `gelf-compression-type: none` without the risk of breaking users.

Seems like the only option left is to recommend disabling gelf compression in docs.

**Comment by kolyshkin** [comment]
Closing as per previous comment. Will follow up with a doc update PR

**Comment by kolyshkin** [comment]
> Seems like the only option left is to recommend disabling gelf compression in docs.
> Will follow up with a doc update PR

https://github.com/docker/docker.github.io/pull/9881
